File: simple-statemachine/simple-statemachine-java/src/main/java/com/example/statemachine/simple/doument/config/listener/StateMachineListenerAdapterImpl.java
package com.example.statemachine.simple.doument.config.listener;

import com.example.statemachine.simple.doument.DocumentEvent;
import com.example.statemachine.simple.doument.DocumentState;
import java.util.Optional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.statemachine.listener.StateMachineListenerAdapter;
import org.springframework.statemachine.state.State;

public class StateMachineListenerAdapterImpl
  extends StateMachineListenerAdapter<DocumentState, DocumentEvent> {

  private static final Logger logger
    = LoggerFactory.getLogger(StateMachineListenerAdapterImpl.class);

  private static String getStateString(
    final State<DocumentState, DocumentEvent> state) {
    final var optional = Optional.ofNullable(state);

    return optional.isEmpty()
      ? "null" : optional.get().getId().toString();
  }

  @Override
  public void stateChanged(
    State<DocumentState, DocumentEvent> from, State<DocumentState, DocumentEvent> to) {

    final var fromStateString = getStateString(from);
    final var toStateString = getStateString(to);

    logger.info("문서의 상태가 {} 에서 {} 로 변경되었습니다.", fromStateString, toStateString);
  }
}

File: simple-statemachine-with-persistence/using-statemachine-recipe/simple-statemachine-kotlin/src/main/kotlin/com/example/statemachine/simple/document/config/lintener/StateMachineListenerAdapterImpl.kt
package com.example.statemachine.simple.document.config.lintener

import com.example.statemachine.simple.document.DocumentEvent
import com.example.statemachine.simple.document.DocumentState
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.statemachine.listener.StateMachineListenerAdapter
import org.springframework.statemachine.state.State

class StateMachineListenerAdapterImpl :
  StateMachineListenerAdapter<DocumentState, DocumentEvent>() {
  private val logger: Logger = LoggerFactory.getLogger(StateMachineListenerAdapterImpl::class.java)

  override fun stateChanged(
    from: State<DocumentState, DocumentEvent>?,
    to: State<DocumentState, DocumentEvent>?
  ) {
    logger.info("문서의 상태가 {} 에서 {} 로 변경되었습니다.", from?.id, to?.id)
  }
}
